processing/src/main/java/io/druid/segment/DimensionQueryHelper.java

```diff
@@ -0,0 +1,388 @@
+ * @param <ValueSelectorType> The type of the row value selector (e.g. DimensionSelector) for this dimension
+ */
+public interface DimensionQueryHelper<ActualType extends Comparable<ActualType>, RowValuesType, ValueSelectorType extends ColumnValueSelector>
+{
+  /**
+   * Get a typed column value selector (DimensionSelector, LongColumnSelector, etc.) from a ColumnSelectorFactory.
+   * @param dimensionSpec The dimension of the selector
+   * @param columnSelectorFactory Column value selector provider
+   * @return Column value selector for the dimension specified by dimensionSpec.
+   */
+  ValueSelectorType getColumnValueSelector(DimensionSpec dimensionSpec, ColumnSelectorFactory columnSelectorFactory);
+
+
+  /**
+   * Get the size of a row object.
+   *
+   * The type of the row object will depend on the dimension type, e.g.:
+   *
+   * String type -> IndexedInts row object
+   * Long type -> IndexedLongs row object
+   *
+   * @param rowValues The row object to return the size of
+   * @return size of the row object
+   */
+  int getRowSize(RowValuesType rowValues);
+
+
+  /**
+   * Get the cardinality, if possible, from a dimension value selector object.
+   *
+   * The class of the row object will depend on the dimension type, e.g:
+   *
+   * String type -> DimensionSelector
+   * Long type -> LongColumnSelector
+   *
+   * @param valueSelector The dimension value selector object
+   * @return Cardinality of the dimension value selector object, -1 if cardinality is not available.
+   */
+  int getCardinality(ValueSelectorType valueSelector);
+
+
+  // Functions for QueryableIndexStorageAdapter, FilteredAggregatorFactory
+  /**
+   * Create a single value ValueMatcher, used for filtering by QueryableIndexStorageAdapter and FilteredAggregatorFactory.
+   *
+   * @param cursor ColumnSelectorFactory for creating dimension value selectors
+   * @param value Value to match against
+   * @return ValueMatcher that matches on 'value'
+   */
+  ValueMatcher getValueMatcher(ColumnSelectorFactory cursor, ActualType value);
+
+
+  /**
+   * Create a predicate-based ValueMatcher, used for filtering by QueryableIndexStorageAdapter and FilteredAggregatorFactory.
+   *
+   * @param cursor ColumnSelectorFactory for creating dimension value selectors
+   * @param predicateFactory A DruidPredicateFactory that provides the filter predicates to be matched
+   * @return A ValueMatcher that applies the predicate for this DimensionQueryHelper's value type from the predicateFactory
+   */
+  ValueMatcher getValueMatcher(ColumnSelectorFactory cursor, final DruidPredicateFactory predicateFactory);
+
+
+  /**
+   * Used by CardinalityAggregator.
+   *
+   * Retrieve the current row from dimSelector and add the row values to the hasher.
+   *
+   * @param dimSelector Dimension value selector
+   * @param hasher Hasher used for cardinality aggregator calculations
+   */
+  void hashRow(ValueSelectorType dimSelector, Hasher hasher);
```

Comment:
+1, I think it makes sense to have the query engines each have their own type-handling abstractions. The sprawling nature of this interface says to me that it isn't doing a good job of being a shared abstraction.